lib/interceptor/proxy.js

```diff
@@ -0,0 +1,39 @@
+'use strict'
+
+const { InvalidArgumentError } = require('../core/errors')
+const ProxyAgent = require('../dispatcher/proxy-agent')
+const DispatcherBase = require('../dispatcher/dispatcher-base')
+
+class ProxyInterceptor extends DispatcherBase {
+  constructor (dispatcher, opts) {
+    super()
+    this.dispatcher = dispatcher
+    this.agent = new ProxyAgent(opts)
+  }
+
+  dispatch (opts, handler) {
+    return this.agent.dispatch(opts, handler)
+  }
+
+  close () {
+    return this.dispatcher.close().then(() => this.agent.close())
+  }
+}
+
+module.exports = opts => {
+  if (typeof opts === 'string') {
+    opts = { uri: opts }
+  }
+
+  if (!opts || (!opts.uri && !(opts instanceof URL))) {
+    throw new InvalidArgumentError('Proxy opts.uri or instance of URL is mandatory')
+  }
+
+  if (opts.auth && opts.token) {
+    throw new InvalidArgumentError(
+      'opts.auth cannot be used in combination with opts.token'
+    )
+  }
```

Comment:
move those in the constructor

Reply:
SGTM, but then we will lose the immediate feedback, as it will be delayed until used by the `dispatcher.compose` logic.
So if implementers do:
```js
const interceptor = proxyInterceptor() // <- this won't fail, until next line
dispatcher.compose(interceptor) // <- here it will be thrown
```

Reply:
I'm proposing we drop those wrapper functions and just expose Dispatchers. They only add noise.